### ISSUE TITLE
[BUG] Compilation error when shortcutting dunder method

### ISSUE THREAD

**Issue by Bluenix2** [cython/cython#4332]
**Describe the bug**
Sometimes you want to create shortcuts to not repeat yourself. So you do something similar to `__ne__ = __eq__` (or any other dunder method).

**To Reproduce**
Code to reproduce the behaviour:
```cython
cdef class AlwaysTrue:
    """Class that always returns True when checking for equality or inequality"""

    cdef int value

    def __init__(self, int value):
        self.value = value

    def __eq__(self, other):
        return True

    __ne__ = __eq__
```
Compiling this gives the following errors:
```
always_true.c(3005): warning C4013: 'None' undefined; assuming extern returning int
always_true.c(3005): warning C4047: 'return': 'PyObject *' differs in levels of indirection from 'int'
```
Here is the cythonized C code:
```c
static PyObject *__pyx_tp_richcompare_11always_true_AlwaysTrue(PyObject *o1, PyObject *o2, int op) {
  switch (op) {
    case Py_EQ: {
      return __pyx_pw_11always_true_10AlwaysTrue_3__eq__(o1, o2);
    }
    case Py_NE: {
      return None(o1, o2);
    }
    default: {
      return __Pyx_NewRef(Py_NotImplemented);
    }
  }
}
```
The following code should be fit for a test-case:
```cython
cdef class AlwaysTrue:
    """Class that always returns True when checking for equality"""

    cdef int value

    def __init__(self, int value):
        self.value = value

    def __eq__(self, other):
        return True

    __ne__ = __eq__

def test_shortcut_dunder():
    """
    >>> a, b = test_shortcut_dunder()
    >>> a == b
    True
    >>> a != b
    True
    """
    return AlwaysTrue(5), AlwaysTrue(10)
```
**Expected behavior**
I would expect this to work and for Cython to reveal the `__ne__` dunder to Python.

**Environment (please complete the following information):**
 - OS: Windows
 - Python version: 3.9.5
 - Cython version: 3.0.0a9

**Additional context**
In my case I want to shortcut `__ror__` to `__or__`.

**Comment by da-woods** [comment]
I don't think this is ever likely to work. One of the things that `cdef class` does is convert a lot of the dunder methods into ["type slots"](https://docs.python.org/3/c-api/typeobj.html#pytypeobject-slots) defined in C, which can be called directly, bypassing the standard Python dictionary lookup.

The upshot of this is that Cython expects them to be a function, defined in the class. Thus assignment of something else to it will really confuse it.

It should probably complain before you get to the C code stage - the C code it's generating is probably pretty dodgy. Hence I'll mark it as "error reporting" instead of treating it as a missing feature.

**Comment by da-woods** [comment]
It's probably worth adding: support for the `__r[op]__` special methods is a fairly recent addition to Cython 3 (https://cython.readthedocs.io/en/latest/src/userguide/special_methods.html#arithmetic-methods) and it probably isn't entirely complete yet (the arguments aren't typed yet for example).

**Comment by Bluenix2** [comment]
> The upshot of this is that Cython expects them to be a function, defined in the class. Thus assignment of something else to it will really confuse it.

Understandably, that said I do feel like this is a somewhat common thing to do. We assign `__ror__` to a function, would this not be detectable?

Cython already special-cases methods with these known names, is there I downside to doing the same with what appears to be attributes?

**Comment by da-woods** [comment]
> Cython already special-cases methods with these known names, is there I downside to doing the same with what appears to be attributes?

There's a surprising amount of hidden complication if you were to actually try to do it I think. For example a lot of the functions have some of the argument types special-cased (`__len__` is special-cased to return a `Py_ssize_t` for example). Thus if you did `__abs__ = __len__` (not too ridiculous... it's the l0 norm...) you now have a mismatch of return types.

How about if I do:
```
def global_func(a, b):
   ....

cdef class C:
   __len__ = global_func
```
It looks like it should work, but have I now modified `global_func` so that the first argument is typed as `C` and the return value is a C integer? Or possibly made a copy of it with different argument types?

The easiest thing would be to say: "OK - don't create the type slot and instead just do the assignment as a Python assignment". That would probably work some of the time (because Python's default implementation of the type-slot just looks up the attribute). But it would fail for `__or__` and `__ror__` because they share a single type-slot in C. So once you've specialized `__or__` then Cython has taken over the whole thing. But maybe assignment to `__ror__` should block Cython taking over `__or__` and the whole thing should fall back on attribute lookups?

It's the sort of thing that could be made to work in some cases, but just gets quite hard to cover all the cases. And having a partly working version creates the expectation that it'll work generally. Which is why I'm  a bit sceptical.

**Comment by scoder** [comment]
While I assume that `__ne__ = __eq__` is just a meaningless example, I think it makes sense to allow at least assignments of other dunder methods of the same class – as long as both signatures are the same. That would allow aliasing slot methods.

That being said, it would be nice to see an actual use case for this first.
